Libplanet/Net/Swarm.cs

```diff
@@ -944,104 +962,118 @@ private async Task RefreshAllocate()
-                // we can omit this clause if assume every chain shares
-                // same genesis block...
-                else if (!blockChain.Blocks.ContainsKey(branchPoint))
-                {
-                    toSync = new BlockChain<T>(
-                        blockChain.Policy,
-                        blockChain.Store);
-                }
-                else
+            _logger.Debug("Trying to fill up previous blocks...");
+
+            int retry = 3;
+            while (true)
+            {
+                try
                 {
-                    _logger.Debug("Forking needed. trying to fork...");
-                    toSync = blockChain.Fork(branchPoint);
-                    _logger.Debug("Forking complete. ");
+                    await FillBlocksAsync(
+                        peer, synced, stop, cancellationToken);
+                    break;
                 }
-
-                _logger.Debug("Trying to fill up previous blocks...");
-
-                int retry = 3;
-                while (true)
+                catch (Exception e)
                 {
-                    try
+                    if (retry > 0)
                     {
-                        await FillBlocksAsync(
-                            peer,
-                            toSync,
-                            oldest.PreviousHash,
-                            cancellationToken
+                        _logger.Error(
+                            e,
+                            "FillBlockAsync() failed. retrying..."
                         );
-
-                        break;
+                        retry--;
                     }
-                    catch (Exception e)
+                    else
                     {
-                        if (retry > 0)
-                        {
-                            _logger.Error(
-                                e,
-                                "FillBlockAsync() failed. retrying..."
-                            );
-                            retry--;
-                        }
-                        else
-                        {
-                            throw;
-                        }
+                        throw;
                     }
                 }
+            }
+
+            return synced;
+        }
 
+        private async Task AppendBlocksAsync<T>(
+            BlockChain<T> blockChain,
+            Peer peer,
+            List<Block<T>> blocks,
+            CancellationToken cancellationToken
+        )
+            where T : IAction, new()
+        {
+            // We assume that the blocks are sorted in order.
+            Block<T> oldest = blocks.First();
+            Block<T> latest = blocks.Last();
+            Block<T> tip = blockChain.Tip;
+
+            if (tip == null || latest.Index > tip.Index)
+            {
+                _logger.Debug("Trying to fill up previous blocks...");
+                BlockChain<T> toAppend = await SyncExistingBlocksAsync(
```

Comment:
How about naming this `previousBlocks` to be consistent with the logged message?